### ISSUE TITLE
AsSplitQuery()-Behavior causing issues with Azure

### ISSUE THREAD

**Issue by angelaki** [dotnet/efcore#24690]
I'm sorry I cannot fully reproduce this error yet, but as I already described in a just created stackoverflow (https://stackoverflow.com/questions/67169707/azure-sql-queries-with-same-order-returning-different-resultsets) Azure SQL _somehow_ seams to have problem _keeping_ the order for some queries. This sure breaks the logic of your "AsSplitQuery()" so that I needed to remove it completely and live with the performance decrease.

Is this maybe an already known misbehavior? Is there something I can do to make it work? Anyways, I thought you should know this.

Edit: I now know that SQL Server is not guaranteed to keep the order between queries for not unique columns! The order is determined based on indexes, columns etc.. But what to do in these cases? Guess I'm not the last one to fall in this pit. I now append a `.ThenBy(e => e.UniqueIdentifier)` everytime a order is attached. But it's a pretty technical reason causing this issue and will sure confuse several persons out there!

I don't know how complex this could be under the hood, but would it be possible to always order by entities PK **last** if any / no orders are used? Guess it's getting pretty tricky right now but it is a terrible pitfall right now!

**Comment by ajcvickers** [comment]
@angelaki EF Core should be ensuring a unique ordering when generating SQL for a split query, so it sounds like this is a bug. Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by angelaki** [comment]
What should have been the expected behavior? Should it get ordered by Id last by default? I'll try my very best, but isolating this issue seams pretty tricky.

**Comment by smitpatel** [comment]
EF Core split query will add ordering by PK automatically. It is appended after the ordering specified by the user so that user gets results in the order they asked. So Split queries always have a stable and unique-ly identifying ordering. So the issue you are describing doesn't really occur for split query in theory. So you will need to provide a full repro which shows that split query is not working correctly due to aforementioned reason. It is likely that you are running into a different issue.

**Comment by angelaki** [comment]
Sorry, I'm pretty busy these days, but I'll try to isolate the issue next weekend. It's an querable pushed through multiple services, getting dynamicly ordered by expressions etc..

**Comment by ajcvickers** [comment]
**EF Team Triage:** Closing this issue as the requested additional details have not been provided and we have been unable to reproduce it.

*BTW this is a canned response and may have info or details that do not directly apply to this particular issue. While we'd like to spend the time to uniquely address every incoming issue, we get a lot traffic on the EF projects and that is not practical. To ensure we maximize the time we have to work on fixing bugs, implementing new features, etc. we use canned responses for common triage decisions.*

**Comment by angelaki** [comment]
Sorry it took me a while to isolate the issue - it actually was pretty easy. Here you go:

    public class Appointment
    {
        public Guid Id { get; set; }
        public string Name { get; set; }
        public DateTime? Start { get; set; }
        public ICollection<AppointmentParticipant> Participants { get; set; }
    }

    public class AppointmentParticipant
    {
        public Guid UserId { get; set; }
        public Guid AppointmentId { get; set; }
        public Appointment Appointment { get; set; }
    }

    _dbContext.Appointments.Include(p => p.Participants).AsSplitQuery().OrderBy(a => a.Start).Skip(10).Take(10).ToArrayAsync();

generates the following SQL:

```
SELECT [t].[Id], [t].[Name], [t].[Start]
FROM (
    SELECT [a].[Id], [a].[Name], [a].[Start]
    FROM [Appointment] AS [a]
    ORDER BY [a].[Start]
    OFFSET @__p_0 ROWS FETCH NEXT @__p_0 ROWS ONLY
) AS [t]
ORDER BY [t].[Start], [t].[Id]
```

As you can see the **INNER SQL** is sorted wrong! Furthermore I noticed:

- .Take(X).Skip(X) (reverse Order) generates wrong SQLs
- Order **after** .Take(X).Skip(X) generates wrong SQLs

But that's ok for me. Intended behavior?

**Comment by roji** [comment]
@angelaki what exactly is wrong with the inner SQL? It seems to sort by Start as you've requested in the LINQ query, no?

> .Take(X).Skip(X) (reverse Order) generates wrong SQLs

Take/Skip does not do reverse ordering - try using these operators in LINQ to Objects (without EF Core) to see the expected behavior. If you think the EF-generated SQL is wrong with regards to that, please open an issue with the exact SQL you're seeing and why it doesn't seem right.

> Order after .Take(X).Skip(X) generates wrong SQLs

Here too, I'm guessing your expectations may not be accurate. Take/Skip without a *preceding* OrderBy are problematic, since in SQL table/subquery rows have no determinate order; so what actual rows get selected is completely non-deterministic. Once again, if you suspect a real issue, please open a separate issue with a full LINQ query and the generated SQL.

**Comment by angelaki** [comment]
Thank you for the explanation about these two behaviors! 👍

Back to my initial issue:

As you can see the outer SQL is ordered by `Start` **and** `Id` to ensure the data is always returned in the same order. `Id` gets added by some EFCore magic I guess - I never actually sorted by it. Since the Inner Query gets only ordered by `Start` (what is not unique) its order is not deterministic (at least not _enough_, since it changes depending on selected fields [SQL Server optimization etc.]).

Now that I use `.AsSplitQuery()` the second query will try to get all the participants the first query returned. It will use the same subquery, bit since no columns are actually selected this time the order changes and the SplitQuery return the participants from different appointments!

**Comment by roji** [comment]
Reopening to let @smitpatel take a look.

**Comment by smitpatel** [comment]
The SQL generated by EF Core is accurate. We generate OrderBy-Apply Skip/Take and then add ordering we require so result set is pre-defined.

Now the ordering being non-unique is not entirely issue as long as Skip/Take are not crossing duplicate column set. e.g. If you are sorting by City and values are New York, New York, Seattle, Seattle and you do Skip(0).Take(3) then which row with city value Seattle will be picked is non-deterministic. If you do Take(2) then subquery may not have exact ordering but it will select same rows and then outer query will order them uniquely.

If your main result set is non-deterministic because of the combination of OrderBy/Skip/Take specified then you will get whatever results server gets. Neither EF Core can predict that it will be problematic (since it is data dependent), nor EF Core can do anything fix is because it may change expected results and add additional (at times significant) perf.

**Comment by angelaki** [comment]
Yes, I totally agree with you. But in my case `AsSingleQuery` still works while `AsSplitQuery` is broken since the reload (appointment participants) targets different rows!

**Comment by smitpatel** [comment]
When the issue happens only when you run query multiple times then it would only happens when using AsSplitQuery. Again, EF Core cannot do anything here.

**Comment by angelaki** [comment]
No, AsSplitQuery runs two queries! But the 2nd one does not contain the records of the first one! Just like the outer query gets the primary key as order criteria attached, the inner one needs it, too.

Do you need more code to reproduce it? I just skipped pasting the second query getting executed by one EFCore action.

**Comment by smitpatel** [comment]
No we don't need more code. We are saying we cannot do anything here as you, owner of the data, need to make sure that running your query are not generating non-deterministic result set. The issue can also be reproduced by applying same set of OrderBy/Skip/Take and then selecting only certain columns using anonymous type projection. Then your own 2 queries will give different result sets. 

At the end of the day, it is data-dependent non-deterministic query by user and EF Core cannot do anything about it.

**Comment by angelaki** [comment]
But you are also adding order criterias for regular queries, why don't you for SubSelects?

Statements running just fine as `AsSingleQuery` and returning wrong results as `AsSplitQuery` are pretty dangerous / hard to find for coders. And it could be solved by applying the order to the inner query just like you do for the outer.

**Comment by smitpatel** [comment]
> why don't you for SubSelects?

I already written above. First we don't know that user-specified ordering is problematic. Second, adding additional ordering lowers perf. We cannot automatically add ordering and cause perf penalty for everyone for the sake of a non-deterministic query.

**Comment by angelaki** [comment]
Ah, ok. But why do you for the outer query? I'd expect a full order extension or none. But the current behavior is just dangerous in my opinion.

Ordering the inner by my criteria (Start) only but the outer by mine (Start) plus the PK (Id) seams inconsistent.

**Comment by smitpatel** [comment]
We add orderings we need to generate results. And we only add orderings at very outer-level unless there are nested collections being projected. So if your subquery is 4 levels deep inside the same thing doesn't work. There are also multiple issues in this repo filed to remove those orderings. We also have PR in progress to remove last ordering because very last ordering is not needed. So sorry, your argument doesn't hold. At the very least the core issue remains that you have paging operation which generates non-deterministic results. You are free to add additional ordering to make it deterministic.

**Comment by angelaki** [comment]
I agree with you if no automatic ordering is done at all - and if it gets removed i'm fine. Right now 99,99% of the queries are automatically ordered and the very few that really need it (paged split queries) aren't.

Maybe also only paged split queries could get ordered automatically. Right now it seams like an issue - just like @ajcvickers said: _EF Core should be ensuring a unique ordering when generating SQL for a split query, so it sounds like this is a bug_

**Comment by smitpatel** [comment]
> EF Core should be ensuring a unique ordering when generating SQL for a split query

Which we did at outer level. It is subquery results which are different. Subquery is generated based on user provided query.

**Comment by angelaki** [comment]
Sorry, but I don't really get it. Doing it at outer level isn't even necessary, it's more a _nice gimmick_, while not sorting the inner one breaks the functionallity. Speaking in performance terms is sorting the outer one applied in 99,99% of the queries for no reason while sorting the inner one would be necessary.

Anyways, how about a `.ThenByKey()` order extension method? This way the user could easy and dynamically apply a PK ordering to entities he's not aware of?

**Comment by smitpatel** [comment]
- Sorting on outer level is not a gimmick. It is required ordering when streaming results. Expression tree doesn't tell you if it is streaming query or buffering query.
- Sorting in inner level may break the functionality but it is purely data dependent and the query user wrote and not something EF Core ever know if it breaks functionality or not.
- There is no additional method needed. You can use `ThenBy` after order by in the query to sort by PK and EF Core will sort inner subquery with it for you.
- Dynamic query generation also works with ThenBy since the type of results are deterministic & not a random type, PK properties should be known to user. And user can use `EF.Property` call to generate any mapped property access in EF Core linq query.

**Comment by roji** [comment]
@angelaki for sorting at the outer level, there are various issues where this is already discussed - #19571 may help understand why it's needed.

**Comment by angelaki** [comment]
Ok, I got the point. Seams like you had some trouble with these sorting things already. And I agree, it's not a bug, it's just not a feature. Maybe an extension method, opt-in or what ever _could_ be helpful for ordered paged queries - but guess that'd make things too complicated. I agree with you: the user needs to know what he's doing.

**Comment by angelaki** [comment]
Back again, since my solution has proven me wrong: I really love the idea of my services returning IQueryables of entities other users can consume. I cannot expect them to know how the entities need to be ordered for the query to work. They can reorder the entities the way they need them breaking the inner query.

Imho EFCore needs to be able to return a queryable that can be reordered without breaking it's logic (without even noticing it)! I guess an extension method that forces all queries (inner and outer) to be ordered by the entities PK would be a great solution. Sure opt-in.
